lib/Sema/CSSimplify.cpp

```diff
@@ -7661,6 +7662,36 @@ ConstraintSystem::simplifyDefaultClosureTypeConstraint(
     return SolutionKind::Unsolved;
   }
 
+  // If closure cannot find any other viable binding and defaults to inferring
+  // `() -> $T`, we don't allow the binding if it is connected to an autoclosure
+  // result because an autoclosure cannot accept a closure paramenter unless
+  // the result is bound to a function type through contextual type.
+  auto originTypeVar = originType->getAs<TypeVariableType>();
+  if (originTypeVar) {
+    originTypeVar = getRepresentative(originTypeVar);
```

Comment:
This is not what I head in mind actually. I think a better place to do this would be while inferring bindings - infer method would get an `argument conversion` constraint which would have a autoclosure parameter on the right-hand side, that should trigger removal (or not acceptance) of `DefaultClosureType` constraints into defaults. This way we would just never find and attempt an incorrect type for a closure.

Reply:
Ahh sorry... I think this was a missundestanding of "prevent inferring" from my part, sorry for that... 
And thank you for the detailed explanation, I'll get back to this later today after work :)

Reply:
No worries, thanks for looking into this in the first place! There is no rush :)

Reply:
Just a question, is that deterministic? I mean when it gets the argument conversion constraint `Defaults` should already be there?

Reply:
No, you shouldn't assume that, that's why I mentioned removal or not acceptance in my first comment :)

Reply:
Hey @xedin 
Just made the changes, let me know what you think!